### DISCUSSION TITLE
kubevela源码中添加自定义crd后，进行自定义crd的创建和删除能够成功，但是获取crd实例时报错

### DISCUSSION THREAD

**Discussion by lkjhchen2** [kubevela/kubevela#5831] [Q&A]
1. 在/apis/core.oam.dev/v1alpha1目录下新建了一个目录nacos，其中放置了自定义的crd的struct定义实现以及相关register.go和zz_generated.deepcopy.go定义，其中register.go的内容如下：

package v1alpha1

import (
	"k8s.io/apimachinery/pkg/runtime/schema"
	"sigs.k8s.io/controller-runtime/pkg/scheme"
)

const (
	Group   = "nacos.io"
	Version = "v1alpha1"
)
var (
	SchemeGroupVersion = schema.GroupVersion{Group: Group, Version: Version}
	SchemeBuilder      = &scheme.Builder{GroupVersion: SchemeGroupVersion}
	AddToScheme        = SchemeBuilder.AddToScheme
)
func init() {
	SchemeBuilder.Register(&Nacos{}, &NacosList{})
}

2. 在/apis/core.oam.dev/groupversion_info.go中，对SchemeBuilder进行注册，内容如下：
package core_oam_dev

import (
	"k8s.io/apimachinery/pkg/runtime"

	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1"
	apisix "github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1/nacos"
	nacos "github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1/nacos"
	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha2"
	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
)

func init() {
	// Register the types with the Scheme so the resources can map objects to GroupVersionKinds and back
	AddToSchemes = append(AddToSchemes, v1alpha1.SchemeBuilder.AddToScheme, v1alpha2.SchemeBuilder.AddToScheme, v1beta1.SchemeBuilder.AddToScheme, nacos.SchemeBuilder.AddToScheme)
}

// AddToSchemes may be used to add all resources defined in the project to a Scheme
var AddToSchemes runtime.SchemeBuilder

// AddToScheme adds all Resources to the Scheme
func AddToScheme(s *runtime.Scheme) error {
	return AddToSchemes.AddToScheme(s)
}

3. 自定义了对应的crd的nacos访问客户端，关键部分代码如下：
import (
	"context"
	v1alpha1 "github.com/oam-dev/kubevela/apis/core.oam.dev/v1alpha1/nacos"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/types"
	"k8s.io/apimachinery/pkg/watch"
	"k8s.io/client-go/kubernetes/scheme"
	"k8s.io/client-go/rest"
)

type (
	NacosInterface interface {
		List(ctx context.Context, opts metav1.ListOptions) (*v1alpha1.NacosList, error)
		Get(ctx context.Context, name string, options metav1.GetOptions) (*v1alpha1.Nacos, error)
		Create(ctx context.Context, nacos *v1alpha1.Nacos) (*v1alpha1.Nacos, error)
		Watch(ctx context.Context, opts metav1.ListOptions) (watch.Interface, error)
		Update(ctx context.Context, nacos *v1alpha1.Nacos, opts metav1.UpdateOptions) (*v1alpha1.Nacos, error)
		Delete(ctx context.Context, name string, opts metav1.DeleteOptions) error
		Patch(ctx context.Context, name string, pt types.PatchType, data []byte, opts metav1.PatchOptions, subresources ...string) (result *v1alpha1.Nacos, err error)
		// ...
	}
)

type nacosClient struct {
	restClient rest.Interface
	ns         string
}

func (c *nacosClient) List(ctx context.Context, opts metav1.ListOptions) (*v1alpha1.NacosList, error) {
	result := v1alpha1.NacosList{}
	err := c.restClient.
		Get().
		Namespace(c.ns).
		Resource("nacos").
		VersionedParams(&opts, scheme.ParameterCodec).
		Do(ctx).
		Into(&result)

	return &result, err
}

func (c *nacosClient) Get(ctx context.Context, name string, opts metav1.GetOptions) (*v1alpha1.Nacos, error) {
	result := v1alpha1.Nacos{}
	err := c.restClient.
		Get().
		Namespace(c.ns).
		Resource("nacos").
		Name(name).
		VersionedParams(&opts, scheme.ParameterCodec).
		Do(ctx).
		Into(&result)

	return &result, err
}

func (c *nacosClient) Create(ctx context.Context, nacos *v1alpha1.Nacos) (*v1alpha1.Nacos, error) {
	result := v1alpha1.Nacos{}
	err := c.restClient.
		Post().
		Namespace(c.ns).
		Resource("nacos").
		Body(nacos).
		Do(ctx).
		Into(&result)

	return &result, err
}
4. 定义创建这个客户端的方法，实现内容如下：
type NacosV1Alpha1Interface interface {
	Nacos(namespace string) NacosInterface
}

type NacosV1Alpha1Client struct {
	restClient rest.Interface
}

func NewForClusterConfig(str string) (*NacosV1Alpha1Client, error) {
	config, err := clientcmd.RESTConfigFromKubeConfig([]byte(str))
	if err != nil {
		return nil, errors.Wrapf(err, "can't create kubeconfig by given content")
	}
	return NewForConfigObj(config)
}

func NewForConfigObj(config *rest.Config) (*NacosV1Alpha1Client, error) {
	configCopy := *config
	if err := setConfigDefaults(&configCopy); err != nil {
		return nil, err
	}
	client, err := rest.RESTClientFor(&configCopy)
	if err != nil {
		return nil, err
	}
	return &NacosV1Alpha1Client{client}, nil
}

func (c *NacosV1Alpha1Client) Nacos(namespace string) NacosInterface {
	return &nacosClient{
		restClient: c.restClient,
		ns:         namespace,
	}
}

func setConfigDefaults(config *rest.Config) error {
	gv := v1alpha1.SchemeGroupVersion
	config.GroupVersion = &gv
	config.APIPath = "/apis"
	config.NegotiatedSerializer = scheme.Codecs.WithoutConversion()

	if config.UserAgent == "" {
		config.UserAgent = rest.DefaultKubernetesUserAgent()
	}

	return nil
}
5. 现象问题如下：
使用这个客户端进行自定义crd实例的创建和删除能够成功，但是用它来进行get获取创建出来的crd对象时会报错误：
"v1.GetOptions is not suitable for converting to \"nacos.io/v1alpha1\" in scheme \"pkg/runtime/scheme.go:100\"

调用部分代码如下：
// GetInstance和Nacos方法之后可以正常获得步骤4的访问k8s集群的访问客户端，问题出在Get指定namespace下的crd实例，报上述错误
nacos, err := a.GetInstance(ctx, cluster).Nacos(namespace).Get(ctx, insName, v1.GetOptions{})

6. 想请问下，自定义crd在apiserver中使用时除了在上述schema中进行注册添加，是不是还需要在其他的地方进行注册，kubevela才能正常识别使用这个crd

**Comment by Somefive** [comment]
"自定义crd在apiserver中使用时除了在上述schema中进行注册添加" 这里的“在apiserver中使用” apiserver 是指什么？ KubeVela 的 VelaUX apiserver 还是 Kubernetes apiserver 还是说用 aggregated api 自建的 APIserver？

**Reply by lkjhchen2** [reply]
kubevela中Policy Workflow等自定义CRD注册到的应该是kubenetes的apiserver吧，此处是想达到相同的目的，能够被K8S所识别
